app/src/main/java/io/homeassistant/companion/android/util/TLSWebViewClient.kt

```diff
@@ -0,0 +1,119 @@
+            } else {
+                getActivity((context).baseContext)
+            }
+        }
+        return null
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    override fun onReceivedClientCertRequest(view: WebView, request: ClientCertRequest) {
+        // Let the WebViewActivity know the endpoint requires TLS Client Auth
+        isTLSClientAuthNeeded = true
+
+        // Aim to obtain the private key for the whole lifecycle of the WebViewActivity
+        val activity = getActivity(view.context)
+        if (activity == null) return
+
+        var key: PrivateKey?
+        var chain: Array<X509Certificate>?
+
+        // Get the key and the chain (if the user previously chose)
+        runBlocking {
+            key = keyChainRepository.getPrivateKey()
+            chain = keyChainRepository.getCertificateChain()
+        }
+
+        // If the key is available, process the request
+        if (key != null && chain != null) {
+            request.proceed(key, chain)
+        } else {
+            // If not, then the user must be prompt for a key
+            // The whole operation is wrapped in the selectPrivateKey method but caution as it must occurs outside of the main thread
+            // see: https://developer.android.com/reference/android/security/KeyChain#getPrivateKey(android.content.Context,%20java.lang.String)
+            // Also from now on displaying error message on the UI is more tricky (not on main thread)
+
+            val alias = selectClientCert(activity, request.principals)
+            // null if the user denied access to the key
+            if (alias != null) {
+                // The key should be available now
+                runBlocking {
+                    key = keyChainRepository.getPrivateKey()
+                    chain = keyChainRepository.getCertificateChain()
+                }
+
+                // If we got the key and the chain, then proceed with the request
+                if (key != null && chain != null) {
+                    request.proceed(key, chain)
+                }
+            } else {
+                // Either the user didn't choose a key or no key was available
+                hasUserDeniedAccess = true
+                request.proceed(null, null)
+            }
+        }
+    }
+
+    @RequiresApi(Build.VERSION_CODES.M)
+    private fun selectClientCert(activity: Activity, principals: Array<Principal>?): String? {
+        // prompt the user for a key
+        val kcac = MyKeyChainCallbackAlias(activity.applicationContext)
+        KeyChain.choosePrivateKeyAlias(activity, kcac, arrayOf<String>(), principals, null, null)
+
+        // wait on the user to select a key
+        while (!kcac.ready) {
+            Thread.sleep(200)
+        }
+
+        // load the key and the cert chain
+        runBlocking {
+            if (kcac.alias != null) {
+                keyChainRepository.load(activity.applicationContext, kcac.alias!!)
+            }
+        }
```

Comment:
`KeyChain.choosePrivateKeyAlias` has a callback, we should not be looping until it's ready.

Reply:
This should be resolved